src/v/utils/fragmented_vector.h

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "vassert.h"
+
+#include <cstddef>
+#include <vector>
+
+/**
+ * A very very simple fragmented vector that provides random access like a
+ * vector, but does not store its data in contiguous memory.
+ *
+ * There is no reserve method because we allocate a full fragment at a time.
+ * However, after you populate a vector you might want to call shrink to fit if
+ * your fragment is large. A more advanced strategy could allocate capacity up
+ * front which just requires a bit more accounting.
+ *
+ * The iterator implementation works for a few things like std::lower_bound,
+ * upper_bound, distance, etc... see fragmented_vector_test.
+ */
+template<typename T, size_t fragment_size = 8192>
+class fragmented_vector {
+    static_assert(fragment_size % sizeof(T) == 0);
+    static constexpr size_t elems_per_frag = fragment_size / sizeof(T);
+    static_assert(elems_per_frag >= 1);
+
+public:
+    fragmented_vector() noexcept = default;
+    fragmented_vector(const fragmented_vector&) noexcept = delete;
+    fragmented_vector& operator=(const fragmented_vector&) noexcept = delete;
+    fragmented_vector(fragmented_vector&&) noexcept = default;
+    fragmented_vector& operator=(fragmented_vector&&) noexcept = default;
+    ~fragmented_vector() noexcept = default;
+
+    void push_back(T elem) {
+        if (_size == _capacity) {
+            std::vector<T> frag;
+            frag.reserve(elems_per_frag);
+            _frags.push_back(std::move(frag));
+            _capacity += elems_per_frag;
+        }
+        _frags.back().push_back(elem);
+        ++_size;
+    }
+
+    void pop_back() {
+        vassert(_size > 0, "Cannot pop from empty container");
+        _frags.back().pop_back();
+        --_size;
+        if (_frags.back().empty()) {
+            _frags.pop_back();
+            _capacity -= elems_per_frag;
+        }
+    }
+
+    const T& operator[](size_t index) const noexcept {
+        vassert(index < _size, "Index out of range {}/{}", index, _size);
+        auto& frag = _frags.at(index / elems_per_frag);
+        return frag.at(index % elems_per_frag);
+    }
```

Comment:
yeh I think I need a crash course in this stuff. the intention was I wanted no UB--so std::terminate on out-of-range, and then I could get that checking from `at()`. make sense?

Reply:
I thought you might be going for that. Was just clarifying.

Reply:
did you want to push the `at` for debug only? since you are already doing the vassert? just a thought.

Reply:
I think we should leave it like this for a while just in case there happens to be a bug in here some where. checking size only checks part of the relationships.

Reply:
removed noexcept left bounds checking